libs/net/nodenetworkconfigurationpolicy.py

```diff
@@ -0,0 +1,153 @@
+from ocp_resources.node_network_configuration_policy_latest import NodeNetworkConfigurationPolicy as Nncp
+from ocp_resources.resource import Resource, ResourceEditor
+from timeout_sampler import retry
+
+WAIT_FOR_STATUS_TIMEOUT_IN_SEC = 90
+WAIT_FOR_STATUS_INTERVAL_IN_SEC = 5
+
+
+@dataclass
+class IP:
+    dhcp: bool
+    auto_dns: bool = True
+    enabled: bool = False
+
+
+@dataclass
+class IPv4(IP):
+    pass
+
+
+@dataclass
+class IPv6(IP):
+    autoconf: bool = False
+
+
+@dataclass
+class STP:
+    enabled: bool
+
+
+@dataclass
+class BridgeOptions:
+    stp: STP
+
+
+@dataclass
+class Port:
+    name: str
+
+
+@dataclass
+class Bridge:
+    options: BridgeOptions | None = None
+    port: list[Port] | None = None
+    allow_extra_patch_ports: bool | None = None
+
+
+@dataclass
+class Interface:
+    name: str
+    type: str
+    state: str
+    ipv4: IPv4
+    ipv6: IPv6
+    bridge: Bridge | None = None
+
+
+@dataclass
+class DesiredState:
+    interfaces: list[Interface]
+
+
+class NNCPSuccessStatusNotMet(Exception):
+    pass
+
+
+class NodeNetworkConfigurationPolicy(Nncp):
+    """
+    NodeNetworkConfigurationPolicy object.
+    """
+
+    class Conditions:
+        class Reason:
+            SUCCESSFULLY_CONFIGURED = "SuccessfullyConfigured"
+            FAILED_TO_CONFIGURE = "FailedToConfigure"
+            NO_MATCHING_NODE = "NoMatchingNode"
+
+    def __init__(
+        self,
+        name: str,
+        desired_state: DesiredState,
+        node_selector: dict[str, str] | None = None,
+    ):
+        """
+        Create and manage NodeNetworkConfigurationPolicy
+
+        Args:
+            name (str): Name of the NodeNetworkConfigurationPolicy object.
+            desired_state (DesiredState): Desired policy configuration - interface creation, modification or removal.
+            node_selector (dict, optional): A node selector that specifies the nodes to apply the node network
+                configuration policy to.
+        """
+        super().__init__(
+            name=name,
+        )
+        self._desired_state = desired_state
+        self._node_selector = node_selector
+
+    def to_dict(self) -> None:
+        super().to_dict()
+        if not self.kind_dict and not self.yaml_file:
+            self.res.setdefault("spec", {}).update({
+                "desiredState": asdict(self._desired_state, dict_factory=self.__dict_normalization),
+                "nodeSelector": self._node_selector if self._node_selector else {},
+            })
+
+    @staticmethod
+    def __dict_normalization(data: list[tuple[str, Any]]) -> dict[str, Any]:
+        """Filter out none values and converts key characters containing underscores into dashes."""
+        return {key.replace("_", "-"): val for (key, val) in data if val is not None}
+
+    def clean_up(self) -> bool:
+        self._delete_interfaces()
+        self.wait_for_status_success()
+        return super().clean_up()
```

Comment:
@myakove , how do you treat a teardown which may fail?
In this case, the teardown includes an intermediate step, but the question still stands for any other teardown (because the object deletion may timeout or be rejected).

I am guessing this is super rare and it will be noticeable due to the raised exception. However, I do not see what can be done here except maybe a retry or a stop to the test-run.

Reply:
This has been discussed offline: The lib object should let the exception raise and it will be up to the caller to do something special.

In general, there is nothing that can be done with this.
If it fails, we may leave leftovers.
To mitigate this, in a followup, we should randomize the names of the interfaces and the nncp itself, attempting to avoid a future collision even if something bad happened.
Said that, failure on edit or teardown is not seen on execution, so the priority to handle it is not high.